packages/flame/lib/src/effects/controllers/random_effect_controller.dart

```diff
@@ -0,0 +1,132 @@
+import 'dart:math';
+
+import 'duration_effect_controller.dart';
+import 'effect_controller.dart';
+
+/// An [EffectController] that wraps another effect controller [child] and
+/// randomizes its duration after each reset.
+///
+/// This effect controller works best in contexts were it has a chance to be
+/// executed multiple times, such as within a `RepeatedEffectController`, or
+/// `InfiniteEffectController`, etc.
+///
+/// The child's duration is randomized first at construction, and then at each
+/// reset (`setToStart`). Thus, the child has a concrete well-defined duration
+/// at any point in time.
+class RandomEffectController extends EffectController {
+  RandomEffectController(this.child, this.randomGenerator)
+      : assert(!child.isInfinite, 'Child cannot be infinite'),
+        super.empty() {
+    _initializeDuration();
+  }
+
+  /// Factory constructor that uses a random variable uniformly distributed on
+  /// `[min, max)`.
+  factory RandomEffectController.uniform(
+    DurationEffectController child, {
+    required double min,
+    required double max,
+    Random? random,
+  }) {
+    assert(min >= 0, 'Min value cannot be negative: $min');
+    assert(min < max, 'Max value must exceed min: max=$max, min=$min');
+    return RandomEffectController(
+      child,
+      _UniformRandomVariable(min, max, random),
+    );
+  }
+
+  /// Factory constructor that employs a random variable distributed
+  /// exponentially with rate parameter `beta`. The produced random values will
+  /// have the average duration of `beta`.
+  factory RandomEffectController.exponential(
+    DurationEffectController child, {
+    required double beta,
+    Random? random,
+  }) {
+    assert(beta > 0, 'Beta must be positive: $beta');
+    return RandomEffectController(
+      child,
+      _ExponentialRandomVariable(beta, random),
+    );
+  }
+
+  final DurationEffectController child;
+  final RandomVariable randomGenerator;
+
+  @override
+  bool get isInfinite => false;
+
+  @override
+  bool get isRandom => true;
+
+  @override
+  bool get completed => child.completed;
+
+  @override
+  double? get duration => child.duration;
+
+  @override
+  double get progress => child.progress;
+
+  @override
+  double advance(double dt) => child.advance(dt);
+
+  @override
+  double recede(double dt) => child.recede(dt);
+
+  @override
+  void setToEnd() => child.setToEnd();
+
+  @override
+  void setToStart() {
+    child.setToStart();
+    _initializeDuration();
+  }
+
+  void _initializeDuration() {
+    final duration = randomGenerator.nextValue();
+    assert(
+      duration >= 0,
+      'Random generator produced a negative value: $duration',
+    );
+    child.duration = duration;
+  }
+}
+
+/// [RandomVariable] is an object capable of producing random values with the
+/// prescribed distribution function. Each distribution is implemented within
+/// its own derived class.
+abstract class RandomVariable {
```

Comment:
I have a feeling this could be useful in a general context, not only to random effect controller;
For example, on Gravity I introduced a [method similar to Uniform Random Variable](https://github.com/bluefireteam/gravitational_waves/blob/master/game/lib/game/collections.dart#L4)
If so, we could move it to its own file and add factory methods to create both private implementations.
Thoughts?

Reply:
Probably as a follow-up though?